scipy/optimize/jonkervolgenant/jonkervolgenant.cpp

```diff
@@ -0,0 +1,194 @@
+   with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+   contributors may be used to endorse or promote products derived
+   from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+This code implements the Jonker-Volgenant algorithm for the rectangular
+assignment problem:
+
+    R. Jonker and A. Volgenant. A Shortest Augmenting Path Algorithm for
+    Dense and Sparse Linear Assignment Problems. *Computing*, 38:325-340
+    December 1987.
+
+The code is based on the pseudocode described in pages
+1685-1686 of:
+
+    DF Crouse. On implementing 2D rectangular assignment algorithms.
+    IEEE Transactions on Aerospace and Electronic Systems
+    52(4):1679-1696, August 2016
+    doi: 10.1109/TAES.2016.140952
+
+Author: PM Larsen
+*/
+
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+static int
+augmenting_path(int nc, std::vector<double>& cost, std::vector<double>& u,
+                std::vector<double>& v, std::vector<int>& path,
+                std::vector<int>& row4col,
+                std::vector<double>& shortestPathCosts, int i,
+                std::vector<bool>& SR, std::vector<bool>& SC, double* p_minVal)
+{
+    double minVal = 0;
+
+    int num_remaining = nc;
+    std::vector<int> remaining(nc);
+    // std::iota(remaining.begin(), remaining.end(), 0); //needs C++11
+    for (int it = 0; it < nc; it++) {
+        remaining[it] = it;
+    }
+
+    std::fill(SR.begin(), SR.end(), false);
+    std::fill(SC.begin(), SC.end(), false);
+    std::fill(shortestPathCosts.begin(), shortestPathCosts.end(), INFINITY);
+
+    // find shortest augmenting path
+    int sink = -1;
+    while (sink == -1) {
+
+        int index = -1;
+        double lowest = INFINITY;
+
+        for (int it = 0; it < num_remaining; it++) {
+            int j = remaining[it];
+
+            double r = minVal + cost[i * nc + j] - u[i] - v[j];
+            if (r < shortestPathCosts[j]) {
+                path[j] = i;
+                shortestPathCosts[j] = r;
+            }
+
+            if (shortestPathCosts[j] < lowest ||
+                (shortestPathCosts[j] == lowest && row4col[j] == -1)) {
```

Comment:
From the commit messages, it's clear that this exists as a performance improvement in the integral case; since this also doesn't appear in the IEEE review, I think it would be useful to have a comment spelling this out.

Reply:
I have added a comment here.